### ISSUE TITLE
[patch #7769] Write flash fails for AVR910 programmers

### ISSUE THREAD

**Issue by avrdude-admin** [avrdudes/avrdude#660]
Tue 24 Apr 2012 12:42:13 PM UTC

When downloading hex file with AVR910 compatible programmer it increments address by number of bytes not words (programm memory is two bytes wide).

[file #25719: avr910.patch](https://savannah.nongnu.org/patch/download.php?file_id=25719)

This issue was migrated from https://savannah.nongnu.org/patch/?7769

**Comment by avrdude-admin** [comment]
Joerg Wunsch <joerg_wunsch>
Tue 24 Apr 2012 01:33:45 PM UTC

Given that AVR910 has been around for 9 years now, I find
the accusation of AVRDUDE using the wrong addressing a
little strange.  So far, nobody seems to have had problems
with it during all the time.

I'd like to see comments from active AVR910 users before
I'm convinced to change anything in that area.

Please resubmit the patch as a context (or unified context)
diff, i. e. the output of "diff -c" or "diff -u".  Plain
ed(1)-style diffs are error-prone whenever something changes
in the file afterwards.

**Comment by avrdude-admin** [comment]
Fred <magister>
Thu 09 May 2013 03:32:48 PM UTC

Exactly the same bug as in butterfly (AVR109).

The 'A' command must send the address in word offset, it is currently sent in byte offset.

For instance with a 128 bytes buffer, avrdude send:

A 00 00
A 00 80
A 01 00
A 01 80
etc

each command followed by 128 bytes of data.

This is wrong, the correct offset is:

A 00 00
A 00 40
A 00 80
A 00 C0
etc

in butterfly.c in the butterfly_set_addr() function, I added a:

addr = addr >> 1;

and bam, it works (for flash)

You have not detected the problem before because of another bug:

Boot loader support auto-increment, so avrdude years ago was sending a simple "A 00 00" and the bootloader was auto-incrementing the address. Something has changed since revision 1007, the big "Mega-commit to bring in memory tagging."

I checked out revision 1006, bootstrap/configure/make/test everything works fine with AVR109

I checked out revision 1007, bootstrap/configure/make/test
AVR109 fails

The code also check for auto-incrementing command 'a' but does not care by after and always send an 'A' command, it should not, it's not efficient.
